alephium/contracts/token_bridge/token_bridge_for_chain.ral

```diff
@@ -0,0 +1,69 @@
+Contract TokenBridgeForChain(
+    governanceContractId: ByteVec,
+    localChainId: U256,
+    localTokenBridgeId: ByteVec, // local chain token bridge contract id
+    remoteChainId: U256,
+    remoteTokenBridgeId: ByteVec, // remote chain token bridge contract id
+    mut next: U256,
+    mut next1: U256,
+    mut next2: U256,
+    undoneSequenceTemplateId: ByteVec, // used for create contract by copyCreateContract
+    refundAddress: Address,
+    mut sendSequence: U256
+) extends Sequence(next, next1, next2, undoneSequenceTemplateId, refundAddress), Constants(), TokenBridgeConstants() {
+    pub fn nextSendSequence() -> U256 {
+        checkPermission!(callerContractId!() == localTokenBridgeId, ErrorCodes.InvalidCaller)
+        let msgSequence = sendSequence
+        sendSequence = sendSequence + 1
+        return msgSequence
+    }
+
+    fn checkTransfer(vaa: ByteVec) -> (Bool, ByteVec) {
+        let governance = Governance(governanceContractId)
+        let (emitterChainId, targetChainId, emitterAddress, msgSequence, payload) = governance.parseAndVerifyVAA(vaa)
+        assert!(emitterChainId == remoteChainId, ErrorCodes.InvalidEmitChainId)
+        assert!(emitterAddress == remoteTokenBridgeId, ErrorCodes.InvalidEmitAddress)
+        assert!(targetChainId == localChainId, ErrorCodes.InvalidTargetChainId)
+        return checkSequence(msgSequence), payload
+    }
+
+    @using(readonly = true)
+    fn parseCompleteTransfer(payload: ByteVec) -> (U256, ByteVec, U256, Address, U256) {
+        assert!(byteVecSlice!(payload, 0, 1) == PayloadId.Transfer, ErrorCodes.InvalidPayloadId)
+        let amount = u256From32Byte!(byteVecSlice!(payload, 1, 33))
+        let vaaTokenId = byteVecSlice!(payload, 33, 65)
+        let vaaTokenChainId = u256From2Byte!(byteVecSlice!(payload, 65, 67))
+        let toAddress = byteVecSlice!(payload, 67, 99)
+        let normalizedArbiterFee = u256From32Byte!(byteVecSlice!(payload, 99, 131))
+        assert!(size!(payload) == 131, ErrorCodes.InvalidMessageSize)
+        let recipient = byteVecToAddress!(#00 ++ toAddress)
+        return amount, vaaTokenId, vaaTokenChainId, recipient, normalizedArbiterFee
+    }
+
+    @using(preapprovedAssets = true)
+    pub fn completeTransfer(vaa: ByteVec, caller: Address) -> () {
+        let (sequenceReady, payload) = checkTransfer(vaa)
+        if (!sequenceReady) { // too many undone sequences in between, compact those sequences first
+            return
```

Comment:
Question: just curious, for the relayer, how does it know when to re-try?

Reply:
The relayer can check the `Sequence` contract state. If the relayer has a message in which the sequence is `seq`, the relayer can re-try if `sequence.next + 512 > seq`.

Reply:
I guess `sequence.next` grows when relayer retries, right? If it tries enough times it will probably make `sequence.next + 512 > seq`?

Reply:
The sequence grows only if the transaction is successful. Therefore, if `seq > sequence.next + 512` , the relayer can regularly query the sequence contract state. When `sequence.next + 512 > seq`, the message will be executed successfully.